src/cephadm/design_choices.rst

```diff
@@ -0,0 +1,165 @@
+Cephadm Design Docs
+===================
+
+
+**Housekeeping note:**
+
+If you want to see things added to this document, please add it via Github's `suggestion[Suggest Changes]` feature.
+This allows me to just hit the button in order for the comment to end up in the document.
+
+
+Intro
+-----
+
+Current situation:
+
+Cephadm manages all registered hosts. This means that it needs to periodically scrape data from each
+host to identify changes on the host like:
+
+* disk added/removed
+* daemon added/removed
+* host network/firewall etc has changed
+
+
+Currently, cephadm scrapes each host (up to 10 in parallel[0]) every 6 minutes[1], unless a refresh is forced manually.
+
+Refreshes for disks (ceph-volume), daemons (podman/docker), etc, happen in sequence.
+
+While the refresh model isn't ideal in the first place, the execution time vastly increases with the cluster size > 10
+
+This certainly isn't acceptable since ceph itself scales massively and so has the management part.
+
+There are a couple of solutions to this:
+
+
+1. Pre fetching/caching data on the host
+----------------------------------------
+
+This would allow bin/cephadm to autonomously scrape data that can later be retrieved by mgr/cephadm.
+
+i.e. `cephadm ls` would access an underlying (transparent) cache.
+
+This, btw assumes a **pull** model.
+
+There is at least one possible solutions on *how* to do this.
+
+1.1 daemonize bin/cephadm
+--------------------------
+
+A daemon is able to execute scraping tasks more frequently (as opposed to only when asked by mgr/cephadm)
+
+
+Downsides:
+
+We're still `scp'ing` bin/cephadm from the ceph-mgr to the host which is then executed.
+A daemon on the other hand needs to run continuously on the target system. If ceph-mgr (and the containing bin/cephadm
+script) is updated, the daemon needs to be informed of this event and get updated/restarted as well.
+
+
+1.1.1 HTTP:
+___________
+
+There is a implementation by Paul Cuzner [2] that uses HTTP to serve an endpoint to the hosts metadata.
+
+Upside:
+
+* We could utilize a proper REST API to CRUD all of the daemons/disks etc.
+  Currently it feels a bit weird to call a binary from a remote host (that's another topic though)
+
+
+Downside:
+
+* This is quite some shift in architecture.
+* We still use SSH to bootstrap a node (in order to install bin/cephadm for the first time)
```

Comment:
```suggestion
* We still use SSH to bootstrap a node (in order to install bin/cephadm for the first time)
----------------------------------------
		(jmolmo):
	Now we are installing the "agent" cephadm in the most complicated and strange/weird way possible. 
	Taking a look about how this was solved in other places (kubernetes, prometheus, ...).  There is an agent that is explicitly installed in each node by the administrator person. Maybe we should think about change the way we are installing cephadm in new nodes.
----------------------------------------	
```

Reply:
Absolutely. Albeit these topics are related (or at least affecting each other), we should discuss this separately. Would you mind starting a doc similar to this one?